### PR TITLE
Optional nms methods supporting for rcnn

### PR DESCRIPTION
**Pull request by HaiyangPeng**
This PR adds two soft (linear and gaussian) nms methods to reduce FN in the case of dense object detection. For the compatibility, the original nms method is reserved. Users can set the NMS_METHOD in rcnn.cpp to select specific method. The corresponding relationship is shown below:
// NMS_METHOD 
// 0: original nms
// 1: soft-nms (linear)
// 2: soft-nms (gaussian) 
The whole codes have been tested on my own data, any questions are welcome ^-^

### CODE REVIEW

**Comment by wang-xinyu** [comment]
@nengwp @freedenS Can you guys help review this PR?

File: rcnn/CMakeLists.txt [hunk 1]
```diff
@@ -18,15 +18,15 @@ include_directories(${PROJECT_SOURCE_DIR}/include)
 include_directories(/usr/local/cuda/include)
 link_directories(/usr/local/cuda/lib64)
 # tensorrt
-include_directories(/home/jushi/TensorRT-8.2.1.6/include)
-link_directories(/home/jushi/TensorRT-8.2.1.6/lib)
+include_directories(/xxx/TensorRT-7.0.0.11/include)
```
[the remark this answers is not in the document]
Comment (HaiyangPeng) [reply rcnn/CMakeLists.txt:21]: > Does it support TRT 8?

Yes, I just tested the code under TRT 7. The compatibility of TRT 7 and 8 by @nengwp has not been modified.

File: rcnn/CMakeLists.txt [hunk 2]
```diff
@@ -2,13 +2,13 @@ cmake_minimum_required(VERSION 3.1)
 
 project(rcnn)
 
-add_definitions(-std=c++14)
```
Comment (nengwp) [inline rcnn/CMakeLists.txt:5]: C++14, seems to be a must in my environment, otherwise brings more warnings

File: rcnn/CMakeLists.txt [hunk 3]
```diff
@@ -2,13 +2,13 @@ cmake_minimum_required(VERSION 3.1)
...
+add_definitions(-std=c++11)
 
 option(CUDA_USE_STATIC_CUDA_RUNTIME OFF)
-set(CMAKE_CXX_STANDARD 14)
+set(CMAKE_CXX_STANDARD 11)
 set(CMAKE_BUILD_TYPE Debug)
 
-set(CUDA_NVCC_FLAGS ${CUDA_NVCC_FLAGS};--extended-lambda)
+set(CUDA_NVCC_FLAGS ${CUDA_NVCC_FLAGS};--expt-extended-lambda)
```
Comment (nengwp) [inline rcnn/CMakeLists.txt:12]: Why use  _**--expt-extended-lambda**_， It seems to be unstable

File: rcnn/CMakeLists.txt (hunk 1 above)
[the remark this answers is not in the document]
Comment (nengwp) [reply rcnn/CMakeLists.txt:21]: Why del _**link_directories(/home/jushi/TensorRT-8.2.1.6/lib)**_？
**_include  and lib_**  Should be included as an example

File: rcnn/CMakeLists.txt [hunk 4]
```diff
@@ -18,15 +18,15 @@ include_directories(${PROJECT_SOURCE_DIR}/include)
...
+link_directories(/xxx/TensorRT-7.0.0.11/lib)
 
-set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++14 -Wall -Ofast -Wfatal-errors -D_MWAITXINTRIN_H_INCLUDED")
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++11 -Wall -Ofast -Wfatal-errors -D_MWAITXINTRIN_H_INCLUDED")
 
 cuda_add_library(myplugins SHARED ${PROJECT_SOURCE_DIR}/BatchedNms.cu ${PROJECT_SOURCE_DIR}/PredictorDecode.cu ${PROJECT_SOURCE_DIR}/RoiAlign.cu ${PROJECT_SOURCE_DIR}/RpnDecode.cu ${PROJECT_SOURCE_DIR}/RpnNms.cu ${PROJECT_SOURCE_DIR}/MaskRcnnInference.cu)
 target_link_libraries(myplugins nvinfer cudart)
 
-find_package(OpenCV)
+find_package(OpenCV 3.4.2)
```
Comment (nengwp) [inline rcnn/CMakeLists.txt:29]: It does not seem appropriate to specify a specific opencv version

File: rcnn/README.md [hunk 5]
```diff
@@ -21,7 +21,9 @@ TensorRT 8.x is supported and you can use it.
 
 ## Contributors
 
+<a href="https://github.com/nengwp"><img src="https://avatars.githubusercontent.com/u/46739135?v=4" width="40px;" alt=""/></a>
 <a href="https://github.com/nengwp"><img src="https://avatars.githubusercontent.com/u/44516353?s=96&v=4" width="40px;" alt=""/></a>
+<a href="https://github.com/nengwp"><img src="https://avatars.githubusercontent.com/u/26213470?v=4" width="40px;" alt=""/></a>
```
Comment (nengwp) [inline rcnn/README.md:27]: gitbub link？

File: rcnn/rcnn.cpp [hunk 6]
```diff
@@ -9,15 +9,15 @@
 #include "MaskRcnnInferencePlugin.h"
 #include "calibrator.hpp"
 
-#define DEVICE 0
```
Comment (nengwp) [inline rcnn/rcnn.cpp:12]: Should keep the default

File: rcnn/rcnn.cpp [hunk 7]
```diff
@@ -9,15 +9,15 @@
...
+#define DEVICE 1
 #define BATCH_SIZE 1
 #define BACKBONE_RESNETTYPE R50
 // data
 static const std::vector<float> PIXEL_MEAN = { 103.53, 116.28, 123.675 };
 static const std::vector<float> PIXEL_STD = {1.0, 1.0, 1.0};
 static constexpr float MIN_SIZE = 800.0;
 static constexpr float MAX_SIZE = 1333.0;
-static constexpr int NUM_CLASSES = 80;
+static constexpr int NUM_CLASSES = 1;
```
Comment (nengwp) [inline rcnn/rcnn.cpp:20]: Should keep the default

**Review by nengwp — COMMENTED** [review]
Please at least fix the above problems

File: rcnn/CMakeLists.txt (hunk 2 above)
Comment (HaiyangPeng) [reply rcnn/CMakeLists.txt:5]: For me, c++11 is also a must in my env. BTW, this is not a fixed param and can be specified by users. I will change it back to c++14 for your test.

**Review by nengwp — CHANGES_REQUESTED** [review]
Please make changes according to the suggestion, thanks.

**Comment by HaiyangPeng** [comment]
> Please make changes according to the suggestion, thanks.

I have fixed the bugs you mentioned ^-^

**Comment by wang-xinyu** [comment]
Merged, thanks @HaiyangPeng @nengwp

**Pull request merged**